ES-10063 Add multi-project support for more stats APIs

PeteGillinElastic opened the pull request:
This affects the following APIs:
 - `GET _nodes/stats`:
   - For `indices`, it now prefixes the index name with the project ID (in multi-project clusters). Previously, it didn't tell you which project an index was in, and it failed if two projects had the same index name.
   - For `ingest`, it now gets the pipeline and processor stats for all projects, and prefixes the pipeline ID with the project ID (in multi-project clusters). Previously, it only got them for the default project.
 - `GET /_cluster/stats`:
   - For `ingest`, it now aggregates the pipeline and processor stats for all projects. Previously, it only got them for the default project.
 - `GET /_info`:
   - For `ingest`, same as for `GET /_nodes/stats`.

It does *not* add multi-project support for the `repositories` section on `_nodes/stats`, since repositories aren't MP-ready yet AFAIK.

This is done by making `IndicesService.stats()` and `IngestService.stats()` include project IDs in the `NodeIndicesStats` and `IngestStats` objects they return, and making those stats objects incorporate the project IDs when converting to XContent.

The transitive callers of these two methods are rather extensive (including all callers to `NodeService.stats()`, all callers of `TransportNodesStatsAction`, and so on). To ensure the change is safe, the callers were all checked out, and they fall into the following cases:
 - The behaviour change is one of the desired enhancements described above.
 - There is no behaviour change because it was getting node stats but neither `indices` nor `ingest` stats were requested.
 - There is no behaviour change because it was getting `indices` and/or `ingest` stats but only using aggregate values.
 - In `MachineLearningUsageTransportAction` and `TransportGetTrainedModelsStatsAction`, the `IngestStats` returned will return stats from all projects instead of just the default with this change, but they have been changed to filter the non-default project stats out, so this change is a noop there. (These actions are not MP-ready yet.)
 - `MonitoringService` will be affected, but this is the legacy monitoring module which is not in use anywhere that MP is going to be enabled. (If anything, the behaviour is probably improved by this change, as it will now include project IDs, rather than producing ambiguous unqualified results and failing in the case of duplicates.)

PeteGillinElastic committed:
Add multi-project support for more stats APIs

This affects the following APIs:
 - `GET _nodes/stats`:
   - For `indices`, it now prefixes the index name with the project ID (for non-default projects). Previously, it didn't tell you which project an index was in, and it failed if two projects had the same index name.
   - For `ingest`, it now gets the pipeline and processor stats for all projects, and prefixes the pipeline ID with the project ID. Previously, it only got them for the default project.
 - `GET /_cluster/stats`:
   - For `ingest`, it now aggregates the pipeline and processor stats for all projects. Previously, it only got them for the default project.
 - `GET /_info`:
   - For `ingest`, same as for `GET /_nodes/stats`.

This is done by making `IndicesService.stats()` and `IngestService.stats()` include project IDs in the `NodeIndicesStats` and `IngestStats` objects they return, and making those stats objects incorporate the project IDs when converting to XContent.

The transitive callers of these two methods are rather extensive (including all callers to `NodeService.stats()`, all callers of `TransportNodesStatsAction`, and so on). To ensure the change is safe, the callers were all checked out, and they fall into the following cases:
 - The behaviour change is one of the desired enhancements described above.
 - There is no behaviour change because it was getting node stats but neither `indices` nor `ingest` stats were requested.
 - There is no behaviour change because it was getting `indices` and/or `ingest` stats but only using aggregate values.
 - In `MachineLearningUsageTransportAction` and `TransportGetTrainedModelsStatsAction`, the `IngestStats` returned will return stats from all projects instead of just the default with this change, but they have been changed to filter the non-default project stats out, so this change is a noop there. (These actions are not MP-ready yet.)
 - `MonitoringService` will be affected, but this is the legacy monitoring module which is not in use anywhere that MP is going to be enabled. (If anything, the behaviour is probably improved by this change, as it will now include project IDs, rather than producing ambiguous unqualified results and failing in the case of duplicates.)

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

nielsbauman commented on server/src/main/java/org/elasticsearch/ingest/IngestStats.java:
```diff
@@ -234,52 +263,63 @@ static Stats merge(Stats first, Stats second) {
+    public record PipelineStat(ProjectId projectId, String pipelineId, Stats stats, ByteStats byteStats) {
         static List<PipelineStat> merge(List<PipelineStat> first, List<PipelineStat> second) {
-            var totalsPerPipeline = new HashMap<String, PipelineStat>();
+            record MergeKey(ProjectId projectId, String pipelineId) {}
```
TIL you could define records inside methods! Do you have any idea how that compiles? I'm mostly curious if the compiler extracts that record definition somewhere or whether it's going to redefine that record on every invocation of this method.

nielsbauman replied:
No I think this is perfectly fine. Thanks for the explanation!

nielsbauman commented on server/src/main/java/org/elasticsearch/ingest/IngestStats.java:
```diff
@@ -170,7 +186,20 @@ public static IngestStats merge(IngestStats first, IngestStats second) {
     }
 
-    static Map<String, List<ProcessorStat>> merge(Map<String, List<ProcessorStat>> first, Map<String, List<ProcessorStat>> second) {
+    static Map<ProjectId, Map<String, List<ProcessorStat>>> merge(
```
I feel like there's a significantly more efficient way to do this merging in the way it's currently being used... But optimizations are not the focus of the MP work right now and I'm not even sure if any optimizations here are worth it for how often they'll be used.

nielsbauman commented on test/external-modules/multi-project/build.gradle:
```diff
@@ -11,7 +11,7 @@ esplugin {
 dependencies {
   testImplementation project(path: ':test:test-clusters')
-  clusterModules project(':test:external-modules:test-multi-project')
+  clusterModules(project(':test:external-modules:test-multi-project'), project(":modules:ingest-common"))
```
I think the more common syntax is:
```suggestion
  clusterModules project(':test:external-modules:test-multi-project')
  clusterModules project(":modules:ingest-common")
```

nielsbauman commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/datatiers/DataTierUsageFixtures.java:
```diff
@@ -79,7 +79,7 @@ static NodeIndicesStats buildNodeIndicesStats(RoutingNode routingNode, long byte
             indexStats.computeIfAbsent(shardId.getIndex(), k -> new ArrayList<>()).add(shardStats);
         }
-        return new NodeIndicesStats(COMMON_STATS, Map.of(), indexStats, true);
+        return new NodeIndicesStats(COMMON_STATS, Map.of(), indexStats, Map.of(), true);
```
Nit: can we add a (small) comment here why were using `Map.of()` for the `projectsByIndex` parameter?

PeteGillinElastic replied:
Done. Alternatively, I could make it actually use the correct map, to make the test more realistic. (This test isn't MP-ready yet, so it would just map everything to the default project.) WDYT?

nielsbauman replied:
I'm leaning towards your current approach (empty map with a comment). That's the least amount of code/complexity without really impacting coverage.

nielsbauman commented on x-pack/qa/multi-project/core-rest-tests-with-multiple-projects/build.gradle:
Some YAML tests from the `ingest-common` module are muted in this file because they depend on stats APIs:
```
    '^test/ingest/15_info_ingest/*', // uses cluster info API
    '^test/ingest/70_bulk/Test bulk request with default pipeline', // uses stats API
    '^test/ingest/70_bulk/Test bulk request without default pipeline', // uses stats API
```
However, because we decided to prefix the project ID, we won't be able to unmute those tests... I don't immediately have a good suggestion, but I thought I'd mention it so we're aware. If we decide that that is OK, we should move these lines to the bottom so they're next to the comment explaining why some stats API tests are muted in MP.

nielsbauman commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -564,6 +572,13 @@ IndexShardStats indexShardStats(final IndicesService indicesService, final Index
     }
 
+    private Map<Index, ProjectId> projectsByIndex() {
+        return indices.values()
+            .stream()
+            .map(AbstractIndexComponent::index)
+            .collect(Collectors.toMap(index -> index, index -> clusterService.state().metadata().projectFor(index).id()));
```
I'm wondering if it's worth it to build this map manually instead of using Streams, for performance reasons. I believe the stats API is called pretty regularly by monitoring agents (but I could be mistaken), and if we're doing this for all indices in all projects, the difference between Streams and non-streams might be noticeable. What do you think?

nielsbauman commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -564,6 +572,13 @@ IndexShardStats indexShardStats(final IndicesService indicesService, final Index
     }
 
+    private Map<Index, ProjectId> projectsByIndex() {
+        return indices.values()
+            .stream()
+            .map(AbstractIndexComponent::index)
+            .collect(Collectors.toMap(index -> index, index -> clusterService.state().metadata().projectFor(index).id()));
```
Thinking out loud: by using `Metadata#projectFor`, we throw an exception if the index is not found. I'm not familiar enough with the `IndicesService` to know whether there is a possibility of the service being "behind" the state in `ClusterService`. For instance, when an index is deleted from the cluster state, is the indices service updated atomically? Or if an index is added, is it first added to the cluster state or the `IndicesService`? If you don't already have an answer to those questions (which is completely fine), I think it's worth having a look to be sure what kind of behavior we could see here.

PeteGillinElastic replied:
> (We already have the possibility that the map is empty because it came from an older node, although at least that situation is only temporary.)

N.B. If we go with the thing I called option 3 in the [other thread](https://github.com/elastic/elasticsearch/pull/127650#discussion_r2076061457), this comment will not be true. The _only_ way that I think this the index could be missing from the map is if we hit this rare timing issue. So I think the question becomes: Do we want to do the plumbing to make an `IndexService` know its project ID in order to avoid this?

(There is, incidentally, a risk that the stats call could fail if there are two indices with the same name in different projects both of which hit this rare timing issue at the same time, causing a key collision. I'm inclined not to worry about that, as it would be transient, although it could be avoided if we just add a counter to the prefix.)

nielsbauman reviewed (commented):
Thanks a lot for working on this, Pete! I did a first round of review by just looking at the production files. I'll have a look at the tests later on. This feels like the right direction; my comments are mostly small and/or thinking out loud.

samxbr reviewed (commented):
LGTM overall! I haven't gone over the tests yet, will run another pass for the tests tomorrow.

PeteGillinElastic committed:
Update test/external-modules/multi-project/build.gradle

Change suggested by Niels.

Co-authored-by: Niels Bauman <33722607+nielsbauman@users.noreply.github.com>

PeteGillinElastic committed:
Respond to review comments

PeteGillinElastic reviewed (commented):
Thanks. I just pushed a commit which responds to most of the comments here. There's some discussion below.

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into ES-10063-mp-stats

PeteGillinElastic committed:
fix merge weirdness

Commit:
[CI] Auto commit changes from spotless

PeteGillinElastic commented on server/src/main/java/org/elasticsearch/indices/NodeIndicesStats.java:
```diff
@@ -291,6 +314,17 @@ public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params outerP
+        ProjectId projectId = projectsByIndex.get(index);
+        if (Objects.equals(projectId, Metadata.DEFAULT_PROJECT_ID)) {
+            return index.getName();
+        }
```
On reflection, I think I'm leaning towards option 3.

This would also mean that, when deserializing an instance from an older node, we could make the map be null and so not show project prefixes: I think this would be valid, because I assume that we will never have to support BWC with nodes older than the current `main` in MP clusters (there's no way that could possibly work!).

PeteGillinElastic replied:
I just pushed a commit switching to option 3. I know you said to park it, but it's an easy change, and I think it might help the other discussion (esp. for folks who haven't seen the PR before we raise it) if we're looking at the closest we have to a proposal.

PeteGillinElastic replied:
Oh, ugh, it now occurs to me: `IngestStats` still behaves the old way (prefix project ID if non-default). They should be consistent. Switching `IngestStats` to behave the new way looks like a little more work because there's no obvious way to indicate to it whether it's in MP mode or not, we'd have to explicitly pass that it.

PeteGillinElastic replied:
I'm wondering whether XContent params might be an option (for both). I've not used those so I don't know what limitations there might be.

PeteGillinElastic committed:
Fix test compilation following upstream change to base class

nielsbauman commented on server/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java:
```diff
@@ -709,37 +710,45 @@ static class IngestStats implements ToXContentFragment {
+                            String,
+                            List<org.elasticsearch.ingest.IngestStats.ProcessorStat>> processorStats : processorStatsForProject.getValue()
+                                .entrySet()) {
+                            pipelineIdsByProject.computeIfAbsent(projectId, k -> new HashSet<>()).add(processorStats.getKey());
```
Ahh yeah you're right. I somehow thought the outer loop was over projects. Then this is indeed the only way to do it :+1:.

nielsbauman commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -564,6 +571,15 @@ IndexShardStats indexShardStats(final IndicesService indicesService, final Index
     }
 
+    private Map<Index, ProjectId> projectsByIndex() {
+        Map<Index, ProjectId> map = new HashMap<>();
```
Nit:
```suggestion
        Map<Index, ProjectId> map = new HashMap<>(indices.size());
```

PeteGillinElastic replied:
For some reason github won't let me commit your suggestion — it says "Sorry, the changes couldn't be applied. Refresh and try again." repeatedly — so I'm making this change myself.

nielsbauman commented on server/src/main/java/org/elasticsearch/indices/NodeIndicesStats.java:
```diff
@@ -66,6 +70,7 @@ public class NodeIndicesStats implements Writeable, ChunkedToXContent {
     private final CommonStats stats;
     private final Map<Index, List<IndexShardStats>> statsByShard;
     private final Map<Index, CommonStats> statsByIndex;
+    private final Map<Index, ProjectId> projectsByIndex;
```
It's a shame we have to pass in this map ourselves (and (de)serialize it), as it's pretty much just the `Metadata.ProjectLookup`... It looks like we only need it for the XContent serialization, but I can't think of a way to pass the `ProjectLookup` to that serialization method. I do think it's worth thinking about that some more (and maybe asking others for suggestions). This is going to result in a lot more serialization over the wire.

PeteGillinElastic replied:
I don't think it's a _lot_ more. It's adding one `ProjectId` where there's already a `CommonStats` and a `List<IndexShardStats>`. But I agree it's unfortunate to be sharing copies of this around everywhere. I'll start a thread on the channel, as you suggest.

nielsbauman replied:
Oh I just realized I'm mixing things up. I was thinking of the serialization of `IngestStats`. In `NodeIndicesStats`, this map _will_ result in quite a bit more serialization: a whole map where we serialize every index (not sure if it's every index _on the node_ or really _all_ indices, but still) and every project ID.

PeteGillinElastic replied:
FWIW, it's every index on the node. I still think that `ProjectId` is small compared with other stuff we're already storing for each index on the node...

nielsbauman commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -564,6 +571,15 @@ IndexShardStats indexShardStats(final IndicesService indicesService, final Index
+        Map<Index, ProjectId> map = new HashMap<>();
+        for (IndexService indexShards : indices.values()) {
+            Index index = indexShards.index();
+            clusterService.state().metadata().lookupProject(index).ifPresent(project -> map.put(index, project.id()));
```
Do we want to assert and/or log a warning if the project is not present?

nielsbauman replied:
Ah nvm, I see you're explaining this choice in your other comment. Ignore this comment please.

nielsbauman commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/datatiers/DataTierUsageFixtures.java:
```diff
@@ -79,7 +79,13 @@ static NodeIndicesStats buildNodeIndicesStats(RoutingNode routingNode, long byte
+            COMMON_STATS,
+            Map.of(),
+            indexStats,
+            /* projectsByIndex is only used for rendering as XContent, which not needed here */ Map.of(),
```
Teensiest nit ever:
```suggestion
            // projectsByIndex is only used for rendering as XContent, which not needed here
            Map.of(),
```
The syntax you used is not very common here (AFAIK) and it took me longer than I'm willing to admit to realize the parameter came after the comment.

PeteGillinElastic committed:
Update x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/datatiers/DataTierUsageFixtures.java

Co-authored-by: Niels Bauman <33722607+nielsbauman@users.noreply.github.com>

PeteGillinElastic committed:
Make projects-by-index map nullable and omit in single-project; always include project prefix in XContent in multip-project, even if default; also incorporate one other review comment

PeteGillinElastic committed:
Add a TODO

PeteGillinElastic committed:
update IT to reflect changed behaviour

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into ES-10063-mp-stats

PeteGillinElastic committed:
Switch to using XContent.Params to indicate whether it is multi-project or not

PeteGillinElastic commented:
I just pushed a commit to switch it to using `ToXContent.Params` to distinguish between single- and multi-project behaviour, as discussed elsewhere. I will look into changing `NodeIndicesStats` to using the nested structure we talked about next.

PeteGillinElastic committed:
Refactor NodesStatsMultiProjectIT to common up repeated assertions

PeteGillinElastic committed:
Defer use of ProjectIdResolver in REST handlers to keep tests happy

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into ES-10063-mp-stats

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into ES-10063-mp-stats

PeteGillinElastic committed:
Include index UUID in "unknown project" case

PeteGillinElastic committed:
Make the index-to-project map empty rather than null in the BWC deserialization case.

This works out fine, for the reasons given in the comment. As it happens, I'd already forgotten to do the null check in the one place it's actively used.

PeteGillinElastic committed:
remove a TODO that is done, and add a comment

PeteGillinElastic committed:
fix typo

PeteGillinElastic committed:
Get REST YAML tests working with project ID prefix TODO finish this

PeteGillinElastic committed:
As a drive-by, fix and un-suppress one of the health REST tests

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into ES-10063-mp-stats

Commit:
[CI] Auto commit changes from spotless

PeteGillinElastic committed:
TODO ugh

PeteGillinElastic committed:
Experiment with different stashing behaviour

Commit:
[CI] Auto commit changes from spotless

PeteGillinElastic committed:
Try a more sensible stash behaviour for assertions

PeteGillinElastic committed:
clarify comment

PeteGillinElastic committed:
Make checkstyle happy

PeteGillinElastic committed:
Make the way `Assertion` works more consistent, and simplify implementation

PeteGillinElastic commented:
Those last few commits make some MP REST YAML tests work in the cases where the response contains a project ID prefix.

The way this works isn't the most elegant thing in the world, but I think it's okay. There are two parts to it: We populate an entry in the 'stash' for the `_project_id_prefix_`, and we subtly change the way that the subject of YAML assertions are interpreted.

Here's how the subject of YAML assertions were interpreted before:

1. If there were no references, e.g. `items.0.index.executed_pipelines`, it was interpreted as a path to look up in the response.
2. If the whole thing was a simple reference, e.g. `$body`, it was looked up in the stash instead — in this case, it asserts on the stashed value, it is _not_ interpreted as a path to look up in the response.
3. If it contained a simple reference somewhere inside it, e.g. `nodes.$master.ingest.total.count`, the reference was looked up in the stash and substituted, and the result interpreted as a path to look up in the response.
4. If it contained an extended reference, e.g. `abc${foo}xyz`, the reference was looked up in the stash and substitued, and the assertion was done on the result — it was _not_ interpreted as a path to look up in the response.

This final case is surprising: it means that `nodes.$master.ingest.total.count` and `nodes.${master}.ingest.total.count` would be treated differently, with the former being substituted and then interpreted as a path in the response and the latter being substituted and then interpreted as the literal subject of the assertion. I don't think this can have been the intention. No tests seem to rely on it. This changes it so that these are both substituted and then treated as paths.

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into ES-10063-mp-stats

Commit:
[CI] Auto commit changes from spotless

samxbr commented on modules/ingest-common/src/yamlRestTest/resources/rest-api-spec/test/ingest/15_info_ingest.yml:
```diff
@@ -85,22 +85,22 @@ teardown:
+  - match: { "ingest.pipelines.${_project_id_prefix_}ingest_info_pipeline.current": 0 }
+  - match: { "ingest.pipelines.${_project_id_prefix_}ingest_info_pipeline.failed": 0 }
+  - gt: { "ingest.pipelines.${_project_id_prefix_}ingest_info_pipeline.ingested_as_first_pipeline_in_bytes": 0 }
+  - gt: { "ingest.pipelines.${_project_id_prefix_}ingest_info_pipeline.produced_as_first_pipeline_in_bytes": 0 }
```
I am not very familiar, could you help me understand how does the `${...}` part get parsed and resolved to the value in the stash?

samxbr replied:
Makes sense, thanks for the detail explanation!

samxbr commented on x-pack/qa/multi-project/core-rest-tests-with-multiple-projects/build.gradle:
```diff
@@ -40,15 +40,11 @@ tasks.named("yamlRestTest").configure {
     '^indices.shard_stores/*/*',
-    '^test/ingest/15_info_ingest/*', // uses cluster info API
-    '^test/ingest/70_bulk/Test bulk request with default pipeline', // uses stats API
-    '^test/ingest/70_bulk/Test bulk request without default pipeline', // uses stats API
```
Nice, thanks for making these tests work in MP!

samxbr commented on x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/MachineLearningInfoTransportActionTests.java:
```diff
@@ -987,20 +988,23 @@ private Map<String, Integer> setupComplexMocks() {
-                                        ),
-                                        // Adding a non inference processor that should be ignored
-                                        new IngestStats.ProcessorStat("grok", "grok", new IngestStats.Stats(100, 100, 100, 100))
+                                    ProjectId.DEFAULT,
```
We can use `randomProjectIdOrDefault()` to make unit tests a little more generic to MP

PeteGillinElastic replied:
These tests fail for non-default project IDs, sadly. This ML usage stuff isn't MP-ready yet (see e.g. the annotations in `MachineLearningUsageTransportAction`).

We'll presumably need to get that fixed at some point, but it's not part of this PR, so I'm doing the minimal change needed to get the tests compiling and passing.

samxbr reviewed (changes requested):
LGTM overall, I have posted some minor comments, it should be good to merge very soon :)

samxbr commented:
> 2. If the whole thing was a simple reference, e.g. `$body`, it was looked up in the stash instead — in this case, it asserts on the stashed value, it is _not_ interpreted as a path to look up in the response.

This makes sense, but it does feel like a little different behaviour from the other 3 cases. I will probably confuse myself on this later, is this behaviour documented somewhere in the code base?

PeteGillinElastic committed:
In RestNodesStatsAction, make the XContent params to channel.request(), which is the value it would have had before this change

PeteGillinElastic reviewed (commented):
Thanks Sam! I hope I replied to everything.

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into ES-10063-mp-stats

PeteGillinElastic commented:
> > 2. If the whole thing was a simple reference, e.g. `$body`, it was looked up in the stash instead — in this case, it asserts on the stashed value, it is _not_ interpreted as a path to look up in the response.
> 
> This makes sense, but it does feel like a little different behaviour from the other 3 cases. I will probably confuse myself on this later, is this behaviour documented somewhere in the code base?

I agree that this is confusing. If we were designing it from scratch now, I'd be pushing for a less confusing syntax, one which made it more explicit what was going on. However, it's too late to change it now: there are thousands of places in the existing YAML tests which rely on it.

I think my PR is making this _less_ confusing, which is always a good thing! After my PR, the rule is "if the subject of the assertion consists of nothing but a reference then it is expanded and used directly as the subject of the assertion; otherwise, any references are expanded and the result is used as a path to look up in the response". Previously, there was the weirdness that it would behave differently depending on whether you wrote `node.$master.stuff` or `node.${master}.stuff` which is extremely surprising.

I'm not aware of any documentation around any of this stuff. I figured it out by reading the code. It took me a while to get to the bottom of it, 'cos you don't get nice error messages. (Also, using the `${...}` format means you need to put things in quotes or the YAML doesn't parse, and if you forget you get a really unhelpful error message where it just tells you that there were no matching tests.)

samxbr commented on server/src/main/java/org/elasticsearch/cluster/project/ProjectIdResolver.java:
```diff
@@ -27,4 +28,14 @@ public interface ProjectIdResolver {
+     * Otherwise, it should return {@code true} to indicate the cluster can accommodate multiple projects regardless
+     * how many project it current has.
+     */
+    default boolean supportsMultipleProjects() {
```
I see, avoid exposing cluster state does make sense.

samxbr commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesStatsAction.java:
```diff
@@ -179,7 +188,14 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
+            .nodesStats(nodesStatsRequest, new RestRefCountedChunkedToXContentListener<>(channel, xContentParamsForRequest(request)));
+    }
+
+    private ToXContent.DelegatingMapParams xContentParamsForRequest(RestRequest request) {
```
Ah I see, it's about retaining the incoming params, thanks for the explanation!

samxbr commented:
> I think my PR is making this less confusing, which is always a good thing! After my PR, the rule is "if the subject of the assertion consists of nothing but a reference then it is expanded and used directly as the subject of the assertion; otherwise, any references are expanded and the result is used as a path to look up in the response". Previously, there was the weirdness that it would behave differently depending on whether you wrote node.$master.stuff or node.${master}.stuff which is extremely surprising.
I'm not aware of any documentation around any of this stuff. I figured it out by reading the code. It took me a while to get to the bottom of it, 'cos you don't get nice error messages. (Also, using the ${...} format means you need to put things in quotes or the YAML doesn't parse, and if you forget you get a really unhelpful error message where it just tells you that there were no matching tests.)

I just found that we have [this YAML test doc](https://github.com/elastic/elasticsearch/blob/main/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/README.asciidoc#embedded_stash_key) which document some of the behaviour (not as comprehensible as your comment I think). I think you have fixed a bug where `${...}` was not actually doing a path look up :)

samxbr reviewed (approved):
LGTM, and thanks for the detailed answers to my questions! I am sorry that this PR drags for so long.

PeteGillinElastic commented:
> > I think my PR is making this less confusing, which is always a good thing! After my PR, the rule is "if the subject of the assertion consists of nothing but a reference then it is expanded and used directly as the subject of the assertion; otherwise, any references are expanded and the result is used as a path to look up in the response". Previously, there was the weirdness that it would behave differently depending on whether you wrote node.$master.stuff or node.${master}.stuff which is extremely surprising.
> > I'm not aware of any documentation around any of this stuff. I figured it out by reading the code. It took me a while to get to the bottom of it, 'cos you don't get nice error messages. (Also, using the ${...} format means you need to put things in quotes or the YAML doesn't parse, and if you forget you get a really unhelpful error message where it just tells you that there were no matching tests.)
> 
> I just found that we have [this YAML test doc](https://github.com/elastic/elasticsearch/blob/main/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/README.asciidoc#embedded_stash_key) which document some of the behaviour (not as comprehensible as your comment I think). I think you have fixed a bug where `${...}` was not actually doing a path look up :)

Oh, good work finding that documentation. I guess I didn't look deeply enough... Anyway, yes, I think you're right (which seems sensible, the old behaviour didn't really make sense).

PeteGillinElastic commented:
> LGTM, and thanks for the detailed answers to my questions! I am sorry that this PR drags for so long.

Thanks! Glad we got there in the end...

PeteGillinElastic commented:
Oh no, I mashed the button too quickly and failed to tidy up the commit message... oh well...
